### DISCUSSION TITLE
upsertMany Provided List<Json>, expected SimilarCreateWithoutMovieInput issue

### DISCUSSION THREAD

**Discussion by ridhwaans** [prisma/prisma#7890] [Q&A] [answered]
using `"@prisma/client": "^2.25.0"`
```
 on prisma.upsertOneMovie. Provided List<Json>, expected SimilarCreateWithoutMovieInput or SimilarCreateWithoutMovieInput or SimilarUncheckedCreateWithoutMovieInput or SimilarUncheckedCreateWithoutMovieInput:
[0] type SimilarCreateWithoutMovieInput {
[0]   id: Int
[0]   backdrop_path: String
[0]   title: String
[0]   name: String
[0]   release_date: DateTime
[0]   overview: String
[0]   show: TVShowCreateNestedOneWithoutSimilarInput
[0] }
```
code
```ts
const { PrismaClient } = require('@prisma/client')
const prisma = new PrismaClient()
...
        let result = await apiRequest
        let similar = result.similar.results.map(similar_movie => {
          return {
            id: similar_movie.id, 
            backdrop_path: similar_movie.backdrop_path, 
            title: similar_movie.title,
            name: similar_movie.name,
            release_date: similar_movie.release_date,
            overview: similar_movie.overview,
            movieId: result.id
          }
        })

        const movie = {
        id: result.id,
...
        similar: {
            upsert: similar
          }
        }
        const upsertMovie = await prisma.movie.upsert({
          where: { id: movie.id },
          update: movie,
          create: movie,
        })
```

here is the `schema.prisma`
```prisma
model Movie {
  id      				Int      	@id
...other fields and attributes...
similar				Similar[]
}

model Similar {
	id				Int			@id
	backdrop_path	String
	title			        String
	name			String
	release_date	        String
	overview		        String
	movie			Movie 		@relation(fields: [movieId], references: [id])
	show			TVShow 		@relation(fields: [showId], references: [id])
	movieId			Int
	showId			Int
}
```

`Similar` is supposed to be an array `[]` of other `Movies` nested in a specific `Movie` object, cant be itself  

I do not have experience with prisma upsert on generated types and I am getting the above error. Im expecting to upsert a `Movie` and at the same time upsert multiple records of `Similar` that are related to `Movie`.   
how do I achieve that?

resources https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-writes

**Comment by ryands17** [comment, marked as answer]
@ridhwaans 👋 

`upsert` in `similar` requires `create`, `where`, and `update but that's not provided. Also it seems that `showId` is required but not passed in `Similar` which could also cause an issue.

**Reply by ridhwaans** [reply]
I have changed the query to the following 
```
const movie = {
        id: result.id,
...
       similar: { 
            updateMany: {
              data: similar
            }
          }
        }
        const upsertMovie = await prisma.movie.upsert({
          where: { id: result.id },
          update: movie,
          create: movie,
        })
```

however, I am getting the error 
```
TypeError: value.map is not a function
[0]     at tryInferArgs (/home/ridhwaans/Source/homehost/server/node_modules/@prisma/client/runtime/index.js:33276:18)
[0]     at valueToArg (/home/ridhwaans/Source/homehost/server/node_modules/@prisma/client/runtime/index.js:33096:16)
[0]     at /home/ridhwaans/Source/homehost/server/node_modules/@prisma/client/runtime/index.js:33341:17
[0]     at Array.reduce (<anonymous>)
[0]     at objectToArgs (/home/ridhwaans/Source/homehost/server/node_modules/@prisma/client/runtime/index.js:33318:28)
[0]     at tryInferArgs (/home/ridhwaans/Source/homehost/server/node_modules/@prisma/client/runtime/index.js:33223:40)
[0]     at valueToArg (/home/ridhwaans/Source/homehost/server/node_modules/@prisma/client/runtime/index.js:33096:16)
[0]     at /home/ridhwaans/Source/homehost/server/node_modules/@prisma/client/runtime/index.js:33341:17
[0]     at Array.reduce (<anonymous>)
[0]     at objectToArgs (/home/ridhwaans/Source/homehost/server/node_modules/@prisma/client/runtime/index.js:33318:28) {
[0]   clientVersion: '2.25.0'
```

`similar` is a javascript type array of a size of 4 objects. the `value.map` error is coming from prisma

**Reply by ridhwaans** [reply]
note: i changed the `Similar` schema to include optional fields 
```
model Similar {
	id				Int			@id @default(autoincrement())
	tmdb_id			Int			@unique
	backdrop_path	String
	title			String?
	name			String?
	release_date	String
	overview		String
	movie			Movie? 		@relation(fields: [movie_id], references: [id])
	show			TVShow? 		@relation(fields: [show_id], references: [id])
	movie_id	Int?
	show_id	Int?
}
```
it is still not creating multiple records of `similar` while creating one record of `movie` as expected  
i have tried `connectOrCreate` but it does not support createmany or upsertmany

**Reply by ryands17** [reply]
Could you share the repo with the code? I can send you the correct code.

Also let me know what exactly do you want to add.

**Reply by ridhwaans** [reply]
@ryands17 yes  
repo is `https://github.com/ridhwaans/homehost/tree/prisma` 

`schema.prisma` is included. database url is `'file:./data/media.db'`

checkout `prisma` branch, line `160` to line `273` `https://github.com/ridhwaans/homehost/blob/prisma/server/jobs/index.js#L160`  

`movie` has 1-to-many relation with `genre[]`, `production_company[]`, `credit[]` and `similar[]` . key is `id` or `tmdb_id`  

for input, you can use json https://homehost-demo.herokuapp.com/api/movies/389 as `result` data.   
what i want is when one `movie` relation is upserted into `movie` table, it needs to correctly also upsert and connect multiple records of `genres`, `production_companies`, `credits` and `similar` into four different tables  

later, if `movie` is deleted, it needs to cascade disconnect corresponding records from its nested relations, or cascade delete them if no `movie` records found for `genre`, `production_company`, `credit` or `similar`

**Reply by ryands17** [reply]
@ridhwaans Solved it! First, there's an issue with the `Similar` model your schema. The `backdrop_path` is required but in the data you shared above, it may also be `null`, so that would cause an error. So your schema should look like:

```prisma
model Similar {
  id            Int     @id @default(autoincrement())
  tmdb_id       Int     @unique
  backdrop_path String?
  title         String?
  name          String?
  release_date  String
  overview      String
  movie         Movie?  @relation(fields: [movie_tmdb_id], references: [tmdb_id])
  show          TVShow? @relation(fields: [show_tmdb_id], references: [tmdb_id])
  movie_tmdb_id Int?
  show_tmdb_id  Int?
}
```

Finally, there were lots of issues with your `movie` object. You cannot use `createMany` in `connectOrCreate`, but you need to pass an array to `connectOrCreate`. This is the final code:

```ts
  let movie = {
    tmdb_id: result.id,
    type: result.type,
    fs_path: result.fs_path,
    url_path: result.url_path,
    ctime: result.ctime,
    mtime: result.mtime,
    adult: result.adult,
    backdrop_path: result.backdrop_path,
    budget: result.budget,
    genres: {
      connectOrCreate: genres.map((g) => ({
        create: g,
        where: { tmdb_id: g.tmdb_id },
      })),
    },
    imdb_id: result.imdb_id,
    overview: result.overview,
    popularity: result.popularity,
    poster_path: result.poster_path,
    production_companies: {
      connectOrCreate: production_companies.map((p) => ({
        create: p,
        where: { tmdb_id: p.tmdb_id },
      })),
    },
    release_date: result.release_date,
    revenue: result.revenue,
    runtime: result.runtime,
    tagline: result.tagline,
    title: result.title,
    vote_average: result.vote_average,
    vote_count: result.vote_count,
    credits: {
      connectOrCreate: credits.map((c) => ({
        create: c,
        where: { tmdb_id: c.tmdb_id },
      })),
    },
    similar: {
      connectOrCreate: similar.map((s) => ({
        create: s,
        where: { tmdb_id: s.tmdb_id },
      })),
    },
  }
  const upsertMovie = await prisma.movie.upsert({
    where: { tmdb_id: result.id },
    update: movie,
    create: movie,
  })
```

This works when I run it separately with the above data.

**Reply by ryands17** [reply]
I would suggest using  <kbd>ctrl + space</kbd> to see all the parameters that you can pass.

**Reply by ridhwaans** [reply]
@ryands17 thank you very much   
that was helpful and pointed me in the right direction   
it's led me to update the schema from 1-to-n to m-n between `Movie` and `Genre`  amongst others     
Im ironing out the last stretch of bugs im seeing where its not growing a nested object as expected of an `upsert` operation   
its difficult to explain because of convoluted data but in essence:     
1) suppose there is a m-n relation between `Seasons` and `Episodes`,   
2) prisma repeatedly receives a unique object with which it creates a new list if it is the first time or it merges with the existing object in prisma. it fails at this part  

let me know if you are interested to discuss or I'd close the issue  
thanks again, appreciate it    


related https://github.com/prisma/prisma/discussions/4339

**Reply by ridhwaans** [reply]
i got it working after splitting the create and update writes for `TV` and `Season`.    
prisma has limitations on recursive upsert and connectOrCreate within each other  
i just had to create once and update and connect the relation separately later. below  
```
const upsertTVShow = await prisma.tVShow.upsert({
        where: { tmdb_id: result.id },
        update: tv_show,
        create: tv_show
      })

for (var s of seasons) {
        const upsertSeason = await prisma.season.upsert({
          where: { tmdb_id: s.tmdb_id },
          update: s,
          create: s
        })
      }
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
